doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+
+## Goals
+
+- Design interface for durable catalog state.
+- Lay the groundwork for splitting out the catalog state from the Coordinator and serving layer.
+- Describe implementations for this interface with the stash and persist.
+- Lay the groundwork for being able to react to catalog changes.
+
+## Non-Goals
+
+- Split out the catalog state from the Coordinator and serving layer.
+- Design a fully differential catalog state that the adapter can react to.
+- Remove the timestamp oracle from the catalog state.
+
+## Overview
+
+The durable catalog interface can currently be found in
+the [`mz_adapter::catalog::storage`](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/storage.rs)
+module. At a high level it stores the following types of objects:
+
+- Audit logs
+- Clusters
+- Cluster replicas
+- Cluster introspection source indexes
+- Config: This is a misc catch all that stores the deployment generation and the stash
+  version. [This issue](https://github.com/MaterializeInc/materialize/issues/20922) describes what
+  it is and how it should be cleaned up.
+- Databases
+- Default privileges
+- ID allocators
+- Items
+- Roles
+- Schemas
+- Settings: This currently only stores the "catalog_content_version" which is subtly different from
+  the version stored in config. They are both used for migrations, the config version is used
+  for stash migrations while the "catalog_content_version" is used for catalog migrations.
+- Storage usage metrics
+- System configurations: This stores system variables (like `max_tables`).
+- System object fingerprints
+- System privileges
+- Timestamps (for timestamp oracles)
+
+We will codify the access and mutation of these objects behind a trait that we will create the
+following implementations for:
+
+- A stash backed implementation.
+- A persist backed implementation.
+- A shadow implementation that uses both the stash and persist implementation for every method and
+  compares results. This will be useful for testing.
+
+A rough prototype of this design can be found
+here (it does not include a persist
+implementation): https://github.com/MaterializeInc/materialize/pull/21071
+
+## Detailed description
+
+### Durable Catalog Trait
+
+We will add the following trait in the catalog to describe interacting with the durable catalog
+state.
+
+NB: `async` modifiers, trait bounds, and most `Result`s are left off for easier readability. Almost
+all methods will return a `Result<T, Error>` for some `T`.
+
+```Rust
+pub trait DurableCatalogState {
+    // Initialization
+
+    /// Reports if the catalog state has been initialized.
+    fn is_initialized(&self) -> bool;
+
+    /// Optionally initialize the catalog if it has not
+    /// been initialized and perform any migrations needed.
+    fn open(&mut self);
+
+    /// Checks to see if opening the catalog would be 
+    /// successful, without making any durable changes.
+    fn check_open(&self) -> Result<(), Error>;
```

Comment:
> If this returns Err, open will also return an Err?

Yes, I've added the `Result` return type to all methods.

> What kinds of things could cause this operation to fail? Migrations not up to date? Not init'd? Other?

I've updated the doc-comment to expand on this. Currently it will return an error if the catalog is not initialized or a migration fails.